src/beamlime/constructors/providers.py

```diff
@@ -0,0 +1,382 @@
+        self, _constructor: Union[Provider, partial, Callable[..., Product]]
+    ) -> None:
+        if isinstance(_constructor, Provider):
+            self._constructor = _constructor.constructor
+        elif isinstance(_constructor, partial):
+            self._constructor = _constructor.func
+        else:
+            self._constructor = _constructor
+
+        _validate_callable_as_provider(self._constructor)
+
+    def _init_arguments(self, _constructor, args, kwargs) -> None:
+        if isinstance(_constructor, (partial, Provider)):
+            self.args = (*_constructor.args, *args)
+            self.keywords = {**_constructor.keywords, **kwargs}
+        else:
+            self.args = args
+            self.keywords = kwargs
+
+    def _init_dependencies(self) -> None:
+        self.arg_dep_specs: DependencySpecDict = collect_argument_specs(
+            self._constructor, *self.args, **self.keywords
+        )
+        self.attr_dep_specs: DependencySpecDict = collect_attr_specs(self._constructor)
+        self.product_spec: ProductSpec = get_product_spec(self._constructor)
+
+    @property
+    def call_name(self) -> str:
+        return ".".join((self.constructor.__module__, self.constructor.__qualname__))
+
+    @property
+    def constructor(self) -> Callable[..., Product]:
+        return self._constructor
+
+    def can_provide(self, product_type: Union[ProductType, ProductSpec]) -> bool:
+        """
+        Check if the given ``product_type`` can be supported by this provider.
+        Product type should be the same type as the returned type of this provider.
+        If they are not the same types,
+        the requested product type should be the parent class
+        of returned type of this provider.
+        """
+        from typing import Any, get_origin
+
+        requested_tp = ProductSpec(product_type).returned_type
+        provided_tp = self.product_spec.returned_type
+
+        if provided_tp in (Any, UnknownType) or requested_tp == provided_tp:
+            return True
+        elif orig_tp := get_origin(requested_tp):
+            # If the requested type is generic.
+            # Note that special form generics such as ``Union``
+            # that are not instance of ``type`` will return ``False``.
+            return self.can_provide(orig_tp)
+        elif isinstance(provided_tp, type) and isinstance(requested_tp, type):
+            # If the requested type is a parent class of the provided type.
+            return issubclass(provided_tp, requested_tp)
+        else:
+            return False
+
+    def __call__(self, *args, **kwargs) -> Product:
+        """Call the constructor with extra arguments."""
+        return self.constructor(*(*self.args, *args), **{**self.keywords, **kwargs})
+
+    def __eq__(self, other: object) -> bool:
+        """Compare ``constructor``, ``args`` and ``keywords`` of the provider."""
+        if not isinstance(other, Provider):
+            raise NotImplementedError(
```

Comment:
Same as before, why not return `False`?

Reply:
It's because the input of ``__setitem__`` is different from the output of ``__getitem__``.
It's because it wraps the item with `Provider` class.
And user might want to do something like below.
So I thought it will be confusing if it just returns False without any explanation...?

```
provider_gr = ProviderGroup()
def foo() -> int:
  return 0

provider_gr[int] = foo
provider_gr[int] == foo  # raise NotImlementedError
```